src/app/modules/submission/IncomingSubmission.class.ts

```diff
@@ -0,0 +1,199 @@
+  }
+
+  /**
+   * Generates a set of response IDs that are verifiable. A response is
+   * said to be verifiable if its corresponding form field has the
+   * isVerifiable property set to `true`.
+   * @returns verifiableResponseIds - Generated set of response IDs.
+   * @protected
+   */
+  private getVerifiableResponseIds(): VerifiableResponseIdSet {
+    return this.responses.reduce<FieldIdSet>((acc, response) => {
+      const responseId = response._id
+      const formField = this.fieldMap[responseId]
+      if (formField.isVerifiable) {
+        acc.add(responseId)
+      }
+      return acc
+    }, new Set()) as VerifiableResponseIdSet
+  }
+
+  /**
+   * Prior to the introduction of IncomingResponse, responses were
+   * stored in a ProcessedFieldResponse[]. This method helps to
+   * create a ProcessedFieldResponse object that can be used in
+   * functions that have not been refactored to take in the new
+   * IncomingResponse object.
+   *
+   * A ProcessedResponse is just a regular FieldResponse with some
+   * metadata inserted into them. This function takes in several
+   * data sets that help determine what each metadata field should
+   * contain.
+   * @param response
+   * @returns processedFieldResponse - The ProcessedFieldResponse
+   * @protected
+   */
+  protected getLegacyProcessedFieldResponse(
+    response: FieldResponse,
+  ): ProcessedFieldResponse {
+    const responseId = response._id
+    const formField = this.fieldMap[responseId]
+    const processedResponse: ProcessedFieldResponse = {
+      ...response,
+      isVisible: this.visibleResponseIds.has(responseId),
+      question: formField.getQuestion(),
+    }
+
+    /**
+     * This pattern of mutating processedResponse is copied from the original
+     * getProcessedResponse function. TODO: Check if can just move this into
+     * the object instantiation stage, preserving the possible undefined value.
+     */
+    if (this.verifiableResponseIds.has(responseId)) {
+      processedResponse.isUserVerified = true
+    }
+
+    return processedResponse
+  }
+
+  protected validate(): Result<true, ProcessingError | ValidateFieldError> {
+    // Guard against invalid form submissions that should have been prevented by
+    // logic.
+    if (
+      getLogicUnitPreventingSubmit(
+        this.responses,
+        this.form,
+        this.visibleFieldIds,
+      )
+    ) {
+      return err(new ProcessingError('Submission prevented by form logic'))
+    }
+
+    const validationResultList = this.responses.map((response) => {
+      const responseId = response._id
+      const formField = this.fieldMap[responseId]
+      return validateField(
+        this.form._id,
+        formField,
+        this.getLegacyProcessedFieldResponse(response),
+      )
+    })
+
+    const validationResultCombined = combine(validationResultList)
+    if (validationResultCombined.isErr()) {
+      return err(validationResultCombined.error)
+    }
```

Comment:
I noticed the behaviour here has changed, since we now validate everything rather than returning early once we find the first invalid answer. since we are doing this anyway, is there value in using `combineWithAllErrors` so we retrieve all the errors?

Reply:
we cant go wrong with being more verbose with our error logging! i just checked that winston logger allows lists of error messages in the `meta` field so this should work out fine

Reply:
realized that `mapRouteError` doesnt expect errors to be lists. Gonna create a generic error-mapper wrapper that takes the first element of an error list to pass into the error-mapper, with appropriate checks for anomaly cases when the list is empty, etc

Reply:
okay! it's not a big deal though so you can also just keep it as is haha